Add option follow for transfer and Rest V2

fredericBregier opened the pull request:
- Option `follow`

This option `follow` allows to add a specific field in the transfer information such that
in an externa tool, it is possible to link multiple transfers as hop of the same file.

For instance, sending a file from host A to host B, which will autoatically resend this file
to host C, and then to host D, all transfers will have a special information within the
transfer information such as `{'follow': number}`. This number will be kept the same
in each and every transfers related to the first one.

This option is enabled by default. Then, for the first transfer, Waarp will allocate
a new specific Id of follow. Then all other transfers related to the first one will
have this one.

If yoy don't want this option enabled, just add `-nofollow` option.

- Add Rest V2 options for FollowId
    
The REST V2 interface integrates the search options based on `followId`
(`GET /v2/transfers/?followId=number`).
    
`number` being possibly a long number, it is adviced to manipulate
it as a String.
    
Add a special finder in DbTaskRunner for this
(`getSelectSameFollowId(followId, orderByStart, limit)`).
    
Fix also Map manipulation within Transfer Information to not loose other informations
than JSON extra information (various usages).

Note, for old versions, it is possible to simulate this option by doing the
following:
- Only for the first transfer, add to the `-info` transfer option a map as
`{"follow": uniqueNumber}` with a unique number such as nanotime or timestamp.
- To request on old versions, one should request all transfers (filtering on date
for instance) and post filtering them on the presence in the `transferInfo` field
of such a Json map.

fredericBregier committed:
Add option follow for transfer

This option `-follow` allows to add a specific field in the transfer information such that
in an externa tool, it is possible to link multiple transfers as hop of the same file.

For instance, sending a file from host A to host B, which will autoatically resend this file
to host C, and then to host D, all transfers will have a special information within the
transfer information such as `{'follow': number}`. This number will be kept the same
in each and every transfers related to the first one.

When specifiying the option `-follow`, for the first transfer, Waarp will allocate
a new specific Id of follow. Then all other transfers related to the first one will
have this one.

fredericBregier committed:
Fix documentation and version

fredericBregier committed:
Add Rest V2 options for FollowId

The REST V2 interface integrates the search options based on `followId`
(`GET /v2/transfers/?followId=number`).

`number` being possibly a long number, it is adviced to manipulate
it as a String.

Add a special finder in DbTaskRunner for this
(`getSelectSameFollowId(followId, orderByStart, limit)`).

Fix also Map manipulation within Transfer Information to not loose other informations
than JSON extra information (various usages).

Note, for old versions, it is possible to simulate this option by doing the
following:
- Only for the first transfer, add to the `-info` transfer option a map as
`{"follow": uniqueNumber}` with a unique number such as nanotime or timestamp.
- To request on old versions, one should request all transfers (filtering on date
for instance) and post filtering them on the presence in the ``transferInfo` field
of such a Json map.

fredericBregier commented:
Note: waiting for new release of Netty to upgrade to last version (corrections of Netty that might impact if not done)

bcarlin commented:
I like the idea of this PR (I have been looking for that for quite some time now).
I'll look at it when the two previous PRs are merged (85% of this PR changes are changes for #48 and #50)...

fredericBregier commented:
@bcarlin Rebase is done, ready for review

bcarlin commented:
thx!

bcarlin commented on doc/waarp-r66/source/configuration-details/detail-tasks.rst:
```diff
@@ -10,7 +87,7 @@ LOG
 Cette tâche loggue ou écrit dans un fichier externe des informations :
 
-- si ``delay`` est 0, aucune sortie ne sera effectuée
+- si delay`` est 0, aucune sortie ne sera effectuée
```
This change breaks the markup!

bcarlin commented on doc/waarp-r66/source/configuration-details/detail-tasks.rst:
```diff
@@ -2,6 +2,83 @@
+    <type>NAME</type>
+    <path>path</path>
+    <delay>x</delay>
+    <rank>n</rank>
```
It is the first time I see this `<rank>` tag...
What does it do?
it is not documented below, and it is not used in any task...

fredericBregier replied:
Added an explanation: this is an optional element (long time ago) which allows, mainly for test and validation purpose, to specify the order of execution, while it is implicit if not specified in written order.

bcarlin commented on doc/waarp-r66/source/changes.rst:
```diff
@@ -7,6 +7,43 @@ La procédure de mise à jour est disponible ici: :any:`upgrade`
+Nouvelles fonctionnalités
+=========================
+
+- [`#49 <https://github.com/waarp/Waarp-All/pull/49>`__]
+  Pour les transferts, une nouvelle option `-follow` permet de gérer le suivi
+  fin des retransferts (rebonds entre plusieurs serveurs R66). Cette option
+  positionne un champ dans la partie `information de transfert` de la forme
+  suivante : `{'follow': numeroUnique}` pour le premier transfert et les
+  transferts suivants récupèreront ainsi cette information nativement.
+- L'interface REST V2 intègre l'option de recherche par `followId`
+  (`GET /v2/transfers/?followId=number`). `number` étant possiblement un entier
+  long, il est conseillé de le manipuler en chaîne de caractères.
+- [`#48 <https://github.com/waarp/Waarp-All/pull/48>`__]
+  Une nouvelle tâche nommée `ICAP` est créée afin de permettre  l'échange avec
+  un serveur répondant à la norme RFC 3507 dite `ICAP`.
+  Elle permet de transférer le contenu du fichier vers un service ICAP via une
+  commande `RESPMOD` et d'obtenir la validation de ce fichier par le service
+  (status `204`).
+
+Correctifs
+==========
+
+- [`#50 <https://github.com/waarp/Waarp-All/pull/50>`__]
+  Le log géré par LogBack génère parfois des logs au démarrage d'information
+  ou de debug qui peuvent être évités (en conservant les Warnings et les Erreurs)
+  via l'ajout dans le fichier de configuration `logback.xml` les paramètres
+  suivants :
```
just fixing markup
```suggestion
- [`#49 <https://github.com/waarp/Waarp-All/pull/49>`__]
  Pour les transferts, une nouvelle option ``-follow`` permet de gérer le suivi
  fin des retransferts (rebonds entre plusieurs serveurs R66). Cette option
  positionne un champ dans la partie ``information de transfert`` de la forme
  suivante : ``{'follow': numeroUnique}`` pour le premier transfert et les
  transferts suivants récupèreront ainsi cette information nativement.
- L'interface REST V2 intègre l'option de recherche par ``followId``
  (``GET /v2/transfers/?followId=number``). ``number`` étant possiblement un entier
  long, il est conseillé de le manipuler en chaîne de caractères.
- [`#48 <https://github.com/waarp/Waarp-All/pull/48>`__]
  Une nouvelle tâche nommée ``ICAP`` est créée afin de permettre  l'échange avec
  un serveur répondant à la norme RFC 3507 dite ``ICAP``.
  Elle permet de transférer le contenu du fichier vers un service ICAP via une
  commande ``RESPMOD`` et d'obtenir la validation de ce fichier par le service
  (status ``204``).

Correctifs
==========

- [`#50 <https://github.com/waarp/Waarp-All/pull/50>`__]
  Le log géré par LogBack génère parfois des logs au démarrage d'information
  ou de debug qui peuvent être évités (en conservant les Warnings et les Erreurs)
  via l'ajout dans le fichier de configuration ``logback.xml`` les paramètres
  suivants :
```

bcarlin commented on doc/waarp-r66/source/changes.rst:
```diff
@@ -7,6 +7,43 @@ La procédure de mise à jour est disponible ici: :any:`upgrade`
+- [`#49 <https://github.com/waarp/Waarp-All/pull/49>`__]
+  Pour les transferts, une nouvelle option `-follow` permet de gérer le suivi
+  fin des retransferts (rebonds entre plusieurs serveurs R66). Cette option
+  positionne un champ dans la partie `information de transfert` de la forme
```
Not a problem *per se*, just a comment :
There is a possibility of confusion in the documentation between "information
de transfert" which is an internal thing to store data about the transfer
(size, reschedule count, now this id, mapped to `runner.fileinfo` in the
database), and the data passed with the ``-info`` argument when the transfert
is started which are sometimes called "Transfer Information", sometimes
"informations de transfert" (mapped to ``runner.transferinfo`` in the DB).

but how to call them, IDK...

fredericBregier replied:
`-info` and  `transfer.transferInfo` and `information de transfert` are indeed the same: they are related to transfer information, such as `FOLLOW` or others like business information. This one is partially under control of Waarp (for extra informations or computations, such as HASH, FOLLOW, RESCHEDULE, include in the Transfer `Map` (JSON Map within the transfer information).

This one is completed by the `path` from the rule dynamically.

The `transfer.fileInfo` is related to the file itself (original name, size), not the transfer itself. This one is out of control of the user (managed by Waarp).

bcarlin commented on WaarpR66/src/main/java/org/waarp/openr66/client/TransferArgs.java:
```diff
@@ -0,0 +1,625 @@
+  public static final String FOLLOW_ARG = "-" + FOLLOW;
+  private static final Option FOLLOW_OPTION =
+      Option.builder(FOLLOW).required(false).hasArg(false)
+            .desc("Specify the transfer should integrate a FOLLOW id").build();
```
```suggestion
            .desc("Specify if the transfer should integrate a FOLLOW id").build();
```

bcarlin reviewed (approved):
Great job !
Once again, I wonder wether this should not be a default setting, reversing the arg `-follow` to `-nofollow`. Would there be a any compatibility with previous versions ?

fredericBregier commented:
Concerning the compatibility with previous version:

- As it is in the `transfer information`, the old versions will include this information but will not able to request it from the database.
- However, an old version will have this information in the database. So it might be possible to request this indirectly (through content of transfer information).
- If the transfer is forward from an old version to another one in the right version, the information will be still valid for the new one (and therefore requestable).
- Moreover, an old version could setup "manually" this option by adding the map `{"follow": number}` in the `-info` option. Of course the number will have to be unique as possible (for instance, using a nanosecond timestamp)

As of setting this as the default: 

- The interest will be to have this by default, even when not needed.
- However, we will have `-nofollow` option which will not backward compatible (as `-follow` option). Such an option (whatever which one) could break the previous versions during argument parsing for old versions.

So, AFAIK, I would suggest to let it as is, and to propose the following:

- For the first transfer, if version is older, set manually within the `-info` the Json MAP
- For the first transfer, if version is new, set the option `-follow`
- For any other transfers (forwarded transfers), adding or not the option changes nothing, whatever new or old versions

So this option is mainly a helper, but it also adds a request on REST V2 for this information. Of course older versions, they can set it manually but they will not be able to request it (except programmatically through `transferInfo` field which is available within REST V1 and V2).

fredericBregier commented:
All fixed, except question on default behavior.

bcarlin commented:
Based on what you said, and a little more thinking on my part, there is no compatibility issue:
- the impact on the protocol is non-existent: it is just a metadata sent with the file (so a client in a pre 3.4.0 version can receive a transfer with a follow id. It won't do anything with it, but the transfer will work)
- the addition of a new command line argument is not an issue (it will be used after an upgrade process, and existing integration will continue to work.

Given that, I would assign a follow id to all new transfer without one already.

> Moreover, an old version could setup "manually" this option by adding the map {"follow": number} in the -info option. Of course the number will have to be unique as possible (for instance, using a nanosecond timestamp)

Are you talking about the `-info` CLI argument given when a transfer is created ?

fredericBregier commented:
Before going to add this as default, yes, it is in the `-info` CLI argument that it can be passed through old client (for the first transfer only, the others will inherit from this one, if they copy (`-copyinfo`) the original arguments.

This is the only information that could be sent to the next transfer, including old versions.

Of course, in order to do that for old versions, one has to change the script usage or crate a fictive option to add this manually (removing this option from the arguments but adding to `-info` the necessary json).

 Now back to the option, as long as this option (`-follow` or `-notfollow`) is not specify for old versions, there is no arm. But when using one new options not declared within the used version, this could break the parsing (unknown option). So the reason to do the right decision.

I let you decide about the default behavior (`-follow` or `-notfollow`).

fredericBregier committed:
Change -follow to -nofollow

Make `-follow` the default behavior. `-nofollow` allows to disable
the creation of follow Id.
Change ordrer of Map and other transfer information within `-info` rebuilt option
such as Map is at the end to not disturb previous usages.
Fix FileInfo vs TransferInfo.

fredericBregier commented:
Fix `option` by default. Let `-nooption` be an excluding option.
Fix the usage of `transferInfo`.
Fix also the `transferInfo` content, such as it is "first" the textual information, then the Json (and not the reverse, which could break previous usages).
@bcarlin You can check

bcarlin commented:
OK, I'll look at it today!

bcarlin commented:
It can be merged !
